src/Groups/MatGrp.jl

```diff
@@ -0,0 +1,808 @@
+import AbstractAlgebra: MatElem, MatSpace, parent_type, Ring, RingElem
+import Hecke: base_ring, det, fq_nmod, FqNmodFiniteField, nrows, tr, trace
+import GAP: FFE
+
+export
+    general_linear_group,
+    matrix_group,
+    MatrixGroup,
+    MatrixGroupElem,
+    omega_group,
+    orthogonal_group,
+    special_linear_group,
+    special_orthogonal_group,
+    special_unitary_group,
+    symplectic_group,
+    unitary_group,
+    GL, GO, GU, SL, SO, Sp, SU
+
+
+
+abstract type AbstractMatrixGroupElem <: GAPGroupElem{GAPGroup} end
+
+"""
+    MatrixGroup{RE<:RingElem, T<:MatElem{RE}} <: GAPGroup
+Type of groups `G` of `n x n` matrices over the ring `R`, where `n = degree(G)` and `R = base_ring(G)`.
+
+At the moment, only rings of type `FqNmodFiniteField` are supported.
+"""
+mutable struct MatrixGroup{RE<:RingElem, T<:MatElem{RE}} <: GAPGroup
+   X::GapObj
+   gens::Vector{<:AbstractMatrixGroupElem}
+   ring::Ring
+   deg::Int
+   descr::Symbol                       # e.g. GL
+   ring_iso::GenRingIso
+   mat_iso::GenMatIso
+
+   function MatrixGroup(m::Int64, F::S) where S<:Ring
+      K = elem_type(S)
+      r = new{K, MatElem{K}}()
+      r.deg = m
+      r.ring = F
+      return r
+   end
+
+   function MatrixGroup(m::Int64, F::S, G_gap::GapObj) where S<:Ring
+      K = elem_type(S)
+      r = new{K, MatElem{K}}()
+      r.deg = m
+      r.ring = F
+      r.X = G_gap
+      return r
+   end
+
+   function MatrixGroup(m::Int64, F::S, L::Vector) where S<:Ring
+      K = elem_type(S)
+      r = new{K, MatElem{K}}()
+      r.deg = m
+      r.ring = F
+      r.gens = L
+      return r
+   end
+end
+
+
+"""
+    MatrixGroupElem{RE<:RingElem, T<:MatElem{RE}} <: GAPGroupElem{MatrixGroup}
+Elements of a group of type `MatrixGroup{RE<:RingElem, T<:MatElem{RE}}`
+"""
+mutable struct MatrixGroupElem{RE<:RingElem, T<:MatElem{RE}} <: AbstractMatrixGroupElem
+   parent::MatrixGroup{RE, T}
+   elm::T
+   X::GapObj
+
+   function MatrixGroupElem(G::MatrixGroup{RE,MatElem{RE}}, x::MatElem{RE}) where RE <: RingElem
+      z = new{RE, MatElem{RE}}()
+      z.parent = G
+      z.elm = x
+      return z
```

Comment:
Out of curiosity, why not simply `return new{RE, MatElem{RE}}(G, x)`  ?

Reply:
I just followed the convention I found on the file `src/HeckeTypes.jl`. I didn't expect the shorter version works, but it does, so I can replace it.

Reply:
Note that calling `new` with arguments can sometimes have a small performance advantage, by not initializing non-bitstype fields twice (a pointer internally). But here it probably doesn't matter at all.